File: elixir/lib/yamlscript.ex
defmodule YAMLScript do
  # Copyright 2023-2026 Ingy dot Net
  # This code is licensed under MIT license (See License for details)

  @moduledoc """
  Elixir binding/API for the libys shared library.

  This module is an Elixir port of the Python 'yamlscript' module,
  the reference implementation for YAMLScript FFI bindings to libys.

  The `load/1` function takes a YAMLScript string as input and
  returns `{:ok, data}` where data is the value the YAMLScript code
  evaluates to, or `{:error, message}`. The `load!/1` variant returns
  the data directly and raises `YAMLScript.Error` on failure.
  """

  @on_load :load_nif

  @doc false
  def load_nif do
    path = :filename.join(:code.priv_dir(:yamlscript), ~c"yamlscript_nif")
    :erlang.load_nif(path, 0)
  end

  @doc """
  Compile and eval a YAMLScript string and return the result.
  """
  @spec load(String.t()) :: {:ok, term()} | {:error, String.t()}
  def load(input) when is_binary(input) do
    case nif_load_ys_to_json(input) do
      {:error, message} ->
        {:error, message}

      json when is_binary(json) ->
        # Decode the JSON response and check for a libys error:
        resp = JSON.decode!(json)

        cond do
          err = resp["error"] ->
            {:error, err["cause"]}

          Map.has_key?(resp, "data") ->
            {:ok, resp["data"]}

          true ->
            {:error, "Unexpected response from 'libys'"}
        end
    end
  end

  @doc """
  Like `load/1` but returns the data directly and raises
  `YAMLScript.Error` on failure.
  """
  @spec load!(String.t()) :: term()
  def load!(input) do
    case load(input) do
      {:ok, data} -> data
      {:error, message} -> raise YAMLScript.Error, message: message
    end
  end

  defp nif_load_ys_to_json(_input) do
    :erlang.nif_error(:nif_not_loaded)
  end
end

defmodule YAMLScript.Error do
  @moduledoc "Error raised by `YAMLScript.load!/1`."
  defexception [:message]
end

File: erlang/src/yamlscript.erl
-module(yamlscript).
-on_load(load_nif/0).

-export([load/1, load_json/1]).
-export([nif_load_json/1]).

load_nif() ->
  Priv = filename:join(filename:dirname(code:which(?MODULE)), "../priv"),
  erlang:load_nif(filename:join(Priv, "yamlscript_nif"), 0).

load(Input) when is_list(Input) ->
  load(list_to_binary(Input));
load(Input) when is_binary(Input) ->
  case load_json(Input) of
    {error, Message} ->
      {error, Message};
    JSON ->
      Resp = json:decode(JSON),
      case maps:get(<<"error">>, Resp, null) of
        null ->
          {ok, maps:get(<<"data">>, Resp)};
        Error ->
          {error, maps:get(<<"cause">>, Error)}
      end
  end.

load_json(Input) when is_list(Input) ->
  load_json(list_to_binary(Input));
load_json(Input) when is_binary(Input) ->
  nif_load_json(Input).

nif_load_json(_Input) ->
  erlang:nif_error(nif_not_loaded).
